code/modules/atmospherics/machinery/components/fusion/hypertorus.dm

```diff
@@ -0,0 +1,1368 @@
+///Mole count required (tritium/hydrogen) to start a fusion reaction
+#define FUSION_MOLE_THRESHOLD				25
+///Used to reduce the gas_power to a more useful amount
+#define INSTABILITY_GAS_POWER_FACTOR 		0.003
+///Used to calculate the toroidal_size for the instability
+#define TOROID_VOLUME_BREAKEVEN				1000
+///Constant used when calculating the chance of emitting a radioactive particle
+#define PARTICLE_CHANCE_CONSTANT 			(-20000000)
+///Conduction of heat inside the fusion reactor
+#define METALLIC_VOID_CONDUCTIVITY			0.15
+///Conduction of heat near the external cooling loop
+#define HIGH_EFFICIENCY_CONDUCTIVITY 		0.95
+///Sets the range of the hallucinations
+#define HALLUCINATION_RANGE(P) 				(min(7, round(abs(P) ** 0.25)))
+///Sets the minimum amount of power the machine uses
+#define MIN_POWER_USAGE						50000
+
+#define DAMAGE_CAP_MULTIPLIER				0.002
+
+//If integrity percent remaining is less than these values, the monitor sets off the relevant alarm.
+#define HYPERTORUS_MELTING_PERCENT 		5
+#define HYPERTORUS_EMERGENCY_PERCENT 	25
+#define HYPERTORUS_DANGER_PERCENT 		50
+#define HYPERTORUS_WARNING_PERCENT 		100
+
+#define WARNING_TIME_DELAY 60
+///to prevent accent sounds from layering
+#define HYPERTORUS_ACCENT_SOUND_MIN_COOLDOWN 3 SECONDS
+
+#define HYPERTORUS_COUNTDOWN_TIME 30 SECONDS
+
+/obj/machinery/atmospherics/components/unary/hypertorus
+	icon = 'icons/obj/atmospherics/components/hypertorus.dmi'
+	icon_state = "core"
+
+	name = "thermomachine"
+	desc = "Heats or cools gas in connected pipes."
+	anchored = TRUE
+	density = TRUE
+	resistance_flags = INDESTRUCTIBLE | LAVA_PROOF | FIRE_PROOF | UNACIDABLE | ACID_PROOF | FREEZE_PROOF
+	layer = OBJ_LAYER
+	pipe_flags = PIPING_ONE_PER_TURF | PIPING_DEFAULT_LAYER_ONLY
+	circuit = /obj/item/circuitboard/machine/thermomachine
+	var/icon_state_open
+	var/icon_state_off
+	var/icon_state_active
+	var/active = FALSE
+	var/fusion_started = FALSE
+
+/obj/machinery/atmospherics/components/unary/hypertorus/Initialize()
+	. = ..()
+	initialize_directions = dir
+
+/obj/machinery/atmospherics/components/unary/hypertorus/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>[src] can be rotated by first opening the panel with a screwdriver and then using a wrench on it.</span>"
+
+/obj/machinery/atmospherics/components/unary/hypertorus/attackby(obj/item/I, mob/user, params)
+	if(!fusion_started)
+		if(default_deconstruction_screwdriver(user, icon_state_open, icon_state_off, I))
+			return
+	if(default_change_direction_wrench(user, I))
+		return
+	if(default_deconstruction_crowbar(I))
+		return
+	return ..()
+
+/obj/machinery/atmospherics/components/unary/hypertorus/getNodeConnects()
+	return list(dir)
+
+/obj/machinery/atmospherics/components/unary/hypertorus/default_change_direction_wrench(mob/user, obj/item/I)
+	. = ..()
+	if(.)
+		SetInitDirections()
+		var/obj/machinery/atmospherics/node = nodes[1]
+		if(node)
+			node.disconnect(src)
+			nodes[1] = null
+		if(length(parents))
```

Comment:
```suggestion
		if(parents[1])
```
Sorry. I test it and found that it really has `parents[1]` position but it null

Reply:
ah so it doesn't matter, no problem!

Reply:
Now it tested and work.
 nullifyPipenet() never remove `parents` list positions. So we need direct check.